### DISCUSSION TITLE
Windows Package Manager 1.2.3411-preview

### DISCUSSION THREAD

**Discussion by denelon** [microsoft/winget-cli#1838] [Releases]
This release represents the 1.2 candidate functionality with experimental features enabled for Windows 10 (1809+), and Windows 11.

We've been busy adding functionality, fixing bugs and documentation.

## Notable Features
* [Revisit the logic in Architecture.cpp for X64 emulation of ARM64 devices.](#666)
* [Require administrator to enable local manifest files.](#1453)
* [Expected return codes](#1278)
* [Installed package tracking](#1614)
* [Added an option to select architecture to install.](#1666)
* [Specify installer architecture](#906)
* [Change download file name to match the installer URL](#1722)
* [Added settings to select preferred/required architecture.](#1727)

## Notable Bug Fix
* [`winget upgrade --all` shouldn't install the same package more than once](#1105)

## What's Changed
* Support for x64 on ARM64 by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1441
* Add colons after agreement labels by @lechacon in https://github.com/microsoft/winget-cli/pull/1440
* Spelling by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1447
* Update spelling.yml by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1448
* Update winget trouble shooting steps  by @ashpatil-msft in https://github.com/microsoft/winget-cli/pull/1454
* Show source identifier with `source list` by @lechacon in https://github.com/microsoft/winget-cli/pull/1455
* Fix crash when resource.pri is not present by @msftrubengu in https://github.com/microsoft/winget-cli/pull/1443
* Disable local manifest by default by @yao-msft in https://github.com/microsoft/winget-cli/pull/1453
* Update localization files by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1462
* doc/Settings.md: Add missing "a" by @me-and in https://github.com/microsoft/winget-cli/pull/1467
* Add expected installer return codes by @lechacon in https://github.com/microsoft/winget-cli/pull/1421
* Remove cancel optimization for packageQuery caller by @sreadingMSFT in https://github.com/microsoft/winget-cli/pull/1472
* Update Docs by @vedantmgoyal2009 in https://github.com/microsoft/winget-cli/pull/1471
* Update Docs [suggestions] by @vedantmgoyal2009 in https://github.com/microsoft/winget-cli/pull/1479
* Minor tweaks to spelling to fix issues by @denelon in https://github.com/microsoft/winget-cli/pull/1480
* Fix Roadmap and Documentation by @vedantmgoyal2009 in https://github.com/microsoft/winget-cli/pull/1468
* Fix Installer return code max value in schema by bad merge by @yao-msft in https://github.com/microsoft/winget-cli/pull/1482
* Source resilience by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1470
* Make list show correlated source even when no upgrade is available by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1481
* Handle single unknown version from REST source by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1483
* Fix debug build failure by @yao-msft in https://github.com/microsoft/winget-cli/pull/1485
* Fix logging of large REST response by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1489
* Create machinery for disabling experimental features by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1490
* Remove motw before calling IAttachmentExecute::Save by @yao-msft in https://github.com/microsoft/winget-cli/pull/1491
* Fix correlation issue by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1492
* Add sample project for calling COM interface by @lechacon in https://github.com/microsoft/winget-cli/pull/1449
* Minor bug fixes by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1494
* Update troubleshooting doc with recent error and findings by @ashpatil-msft in https://github.com/microsoft/winget-cli/pull/1495
* Localize the product name per recommendation by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1498
* Update localization files by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1499
* Make rename retry more frequently for longer, then try making a hardlink by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1497
* Disable test hooks fixes by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1501
* Exit the locale choice loop if the default localization is good enough by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1500
* Ensure available packages match with a single installed package by @lechacon in https://github.com/microsoft/winget-cli/pull/1473
* Updating Readme by @denelon in https://github.com/microsoft/winget-cli/pull/1507
* Add option to check fields requiring verified publisher in manifest p… by @yao-msft in https://github.com/microsoft/winget-cli/pull/1487
* Use correct Subexecution id when working on multiple packages by @ashpatil-msft in https://github.com/microsoft/winget-cli/pull/1504
* Changes for sample COM caller project by @lechacon in https://github.com/microsoft/winget-cli/pull/1514
* Fixed bug where the package name wouldn't be written during upgrade. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1517
* Fix nuget release pipeline by @yao-msft in https://github.com/microsoft/winget-cli/pull/1522
* Fix test task overrides by @lechacon in https://github.com/microsoft/winget-cli/pull/1526
* Fix unit tests in pipeline by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1529
* Add missing resource string by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1532
* Change format to long for installer success codes by @ryfu-msft in https://github.com/microsoft/winget-cli/pull/1533
* Settings streams exchange semantics by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1534
* Added vsconfig file to automatically install required workloads. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1551
* Color contamination fix by @hackean-msft in https://github.com/microsoft/winget-cli/pull/1527
* Added validation of count argument for search/list. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1564
* closes #1569 by @exxmen in https://github.com/microsoft/winget-cli/pull/1580
* Split COM install command into download and install stages by @lechacon in https://github.com/microsoft/winget-cli/pull/1528
* Check enablement state for all output write operations by @hackean-msft in https://github.com/microsoft/winget-cli/pull/1581
* Move Issue Templates to Forms by @Trenly in https://github.com/microsoft/winget-cli/pull/1596
* The source column is no longer shown in the table if a source was specified (list/upgrade). by @jedieaston in https://github.com/microsoft/winget-cli/pull/1598
* Add projected winrt wrapper for server com interface and add tests by @sreadingMSFT in https://github.com/microsoft/winget-cli/pull/1373
* Add link to new discussion for questions by @Trenly in https://github.com/microsoft/winget-cli/pull/1617
* Package type dependency graph specs by @fzanollo in https://github.com/microsoft/winget-cli/pull/1340
* Rearrange choices for ARP changes by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1630
* Introduce Managed Interop and merged manifest classes to WingetUtil nuget  by @jamespik in https://github.com/microsoft/winget-cli/pull/1621
* Update doc with Machine-wide install information by @ashpatil-msft in https://github.com/microsoft/winget-cli/pull/1642
* Added progress indication to InstallMultiplePackages. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1613
* Installed package tracking by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1614
* Package type Dependency graph by @fzanollo in https://github.com/microsoft/winget-cli/pull/1341
* Refactor Source Interface to be more object oriented by @yao-msft in https://github.com/microsoft/winget-cli/pull/1631
* CurrentLocalization is now populated for dependency manifests. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1655
* Fix yaml manifest schema and parsing for ExpectedReturnCodes by @yao-msft in https://github.com/microsoft/winget-cli/pull/1646
* Fix incorrect processing of PackageMatchFields in rest client by @yao-msft in https://github.com/microsoft/winget-cli/pull/1659
* Fix interface casting in client dll. by @sreadingMSFT in https://github.com/microsoft/winget-cli/pull/1673
* Prevent duplicate installations in winget upgrade all by @yao-msft in https://github.com/microsoft/winget-cli/pull/1652
* Use package tracking data to correlate in CompositeSource by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1671
* Add product code and PFN to enable existing searches by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1677
* Added an option to select architecture to install. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1666
* Add preview concept to allow build time determination by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1679
* Print message when install technology is different in upgrade scenarios by @msftrubengu in https://github.com/microsoft/winget-cli/pull/1649
* Prevent circular dependency in global telemetry logger creation by @yao-msft in https://github.com/microsoft/winget-cli/pull/1674
* Made all installer types lowercase during `show`. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1689
* Use SFINAE to stop paths from logging without using u8string by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1697
* Made the Installation Disclaimer for the Microsoft Store bold. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1698
* remove plain by @MagnetarRocket in https://github.com/microsoft/winget-cli/pull/1700
* Make VT enablement/disablement work properly by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1699
* Create shared DLL for winget.exe and COM server by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1687
* Implement parallel downloads for COM scenarios by @lechacon in https://github.com/microsoft/winget-cli/pull/1588
* Updated Valijson to version 0.6. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1721
* Change download file name to match the installer URL by @JohnMcPMS in https://github.com/microsoft/winget-cli/pull/1722
* Updated solution to build under Visual Studio 2022 by @jedieaston in https://github.com/microsoft/winget-cli/pull/1720
* Update component governance manifest with valijson v0.6 commit by @yao-msft in https://github.com/microsoft/winget-cli/pull/1732
* Update source agreement market message to be more clear by @yao-msft in https://github.com/microsoft/winget-cli/pull/1731
* Make telemetry logger per context and send summary event on destruction by @yao-msft in https://github.com/microsoft/winget-cli/pull/1718
* Update Notice from latest component governance output by @yao-msft in https://github.com/microsoft/winget-cli/pull/1755
* Added settings to select preferred/required architecture. by @jedieaston in https://github.com/microsoft/winget-cli/pull/1727
* Update README.md by @denelon in https://github.com/microsoft/winget-cli/pull/1759
* Reduce localized languages per localization team guidance by @yao-msft in https://github.com/microsoft/winget-cli/pull/1761

## New Contributors
* @vedantmgoyal2009 made their first contribution in https://github.com/microsoft/winget-cli/pull/1471
* @exxmen made their first contribution in https://github.com/microsoft/winget-cli/pull/1580
* @MagnetarRocket made their first contribution in https://github.com/microsoft/winget-cli/pull/1700

**Full Changelog**: https://github.com/microsoft/winget-cli/compare/v1.0.12576...v1.2.3411-preview

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/winget-cli/releases/tag/v1.2.3411-preview'>Windows Package Manager 1.2.3411-preview</a>.</em>

**Comment by rcdailey** [comment]
Fantastic work; thank you!

**Comment by doctordns** [comment]
Great stuff.

But where are the PowerShell cmdlets.

This too remains a power toy, which is a shame. Oh well.

**Reply by denelon** [reply]
They are still [work in progress](https://github.com/microsoft/winget-cli/tree/master/tools/PowerShell). We're working on tests and validation.

**Reply by doctordns** [reply]
You should publish them as part of Winget. Sure, publish as preview but please start including these as part of WInget.

**Reply by denelon** [reply]
As soon as we have met the bar for testing and compliance, they will be published with our releases. We will also sign them and publish them to the PowerShell gallery at that point.

**Reply by MrSmithLWL** [reply]
With the powershell enhancements will this also add in support for SSH or WinRM? basically will the reparse be resolvable to the real exe?

**Reply by ghost** [reply]
@MrSmithLWL How those AEA(App Execution Aliases) work under the hood is undocumented. Yes There are reverse engineered Blogs about it. https://www.tiraniddo.dev/2019/09/overview-of-windows-execution-aliases.html

Nevertheless, SSH/WinRM/Docker  each of these corresponding client programs will need to handle AEAs(App Execution Aliases) by implementing the reverse engineered knowledge. But there seems not enough interest to do so atleast for SSH team https://github.com/PowerShell/Win32-OpenSSH/issues/1632
